<!DOCTYPE html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1">
<style>
.accordion {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 1.2em;
  transition: 0.4s;
  font-weight: bold;
}

.active-question, .accordion:hover {
  background-color: #ccc; 
}

.panel {
  padding: 0 18px;
  display: none;
  background-color: white;
  overflow: hidden;
  font-size: 1.1em;
}

.highlight {
  background-color: yellow;
}
</style>
</head>
<body>

<button class="accordion">What is SpliceVault?</button>
<div class="panel">
  <p>SpliceVault houses our resource of the most common mis-splicing events local to the exon-intron junction, based on unannotated splice-junctions detected across 335,301 publicly available RNA-Seq samples from GTEx and SRA, and built in GRCh38 (300K-RNA). It is collectively hosted with an older version of the database, 40K-RNA.</p>
  <p>300K-RNA (top-4 events) is an empirical method which accurately predicts the nature of variant associated mis-splicing with 90% sensitivity</p>
  <p>See <a href="https://doi.org/10.1038/s41588-022-01293-8" target="_blank">Dawes et al. 2022</a> for details on formation of the 300K-RNA database</p>
</div>

<button class="accordion">What's the difference between 300K-RNA & 40K-RNA?</button>
<div class="panel">
  <p>300K-RNA is a database of splice-junctions found across 335,663 publicly available RNA-Seq samples from GTEx and SRA, and built in hg38 (GRCh38). See <a href="https://doi.org/10.1038/s41588-022-01293-8" target="_blank">Dawes et al. 2022</a> for details and methods.</p>
  <p>40K-RNA is an older version of the database, based on 40,233 RNA-Seq samples similarly from GTEx and SRA, but built in hg19 (GRCh37). See <a href="https://doi.org/10.1038/s41588-022-01293-8" target="_blank">Dawes et al. 2022</a> for details and methods. We only recommend using this version of the database if you can't convert your splice site of interest to hg38 for whatever reason.</p>
</div>

<button class="accordion">How was the canonical transcript chosen for each gene?</button>
<div class="panel">
  <strong>GRCh37</strong>
  <ul>
    <li><strong>Ensembl</strong> - Canonical transcript as provided by Ensembl. Ensembl API derives this based on legacy canonical transcript selection rules <sup>*</sup></li>
    <li><strong>Refseq</strong> - Canonical transcript not provided. Calculated using Ensembl’s legacy canonical transcript selection rules <sup>*</sup></li>
  </ul>
  
  <strong>GRCh38</strong>
  <ul>
    <li><strong>Ensembl</strong> - Canonical transcript as provided by Ensembl. Please refer to <a href="http://www.ensembl.org/info/genome/genebuild/canonical.html" target="_blank">http://www.ensembl.org/info/genome/genebuild/canonical.html</a> for further info on how it was derived</li>
    <li><strong>Refseq</strong> - Canonical transcript not provided. Calculated using Ensembl’s legacy canonical transcript selection rules <sup>*</sup></li>
  </ul>
  
  <p><sup>*</sup> The canonical transcript is used in the gene tree analysis in Ensembl and does not necessarily reflect the most biologically relevant transcript of a gene. For human, the canonical transcript for a gene is set according to the following hierarchy:
  <ol>
    <li>Longest CCDS translation with no stop codons.</li>
    <li>If no (1), choose the longest Ensembl/Havana merged translation with no stop codons.</li>
    <li>If no (2), choose the longest translation with no stop codons.</li>
    <li>If no translation, choose the longest non-protein-coding transcript</li>
  </ol>
  (Source: <a href="https://web.archive.org/web/20200928022822/https://m.ensembl.org/Help/Glossary?id=521" target="_blank">https://web.archive.org/web/20200928022822/https://m.ensembl.org/Help/Glossary?id=521</a>)</p>
</div>

<button class="accordion">What are GTEx and SRA/Intropolis?</button>
<div class="panel">
  <p>GTEx provide RNA-Seq data from a large set of healthy individuals. 
SRA (Sequence Read Arcchive) is an online archive of high-throughput RNA sequencing data.
<a href="https://github.com/nellore/intropolis" target="_blank">Intropolis</a> is a set of ~42M exon-exon junctions found across 21,504 human RNA-seq samples from the Sequence Read Archive (SRA).</p>
</div>

<button class="accordion">How do I read the 'Skipped Exons' column?</button>
<div class="panel">
  <p>The column reports the ids of the exons skipped, for example '2' refers to skipping of exon 2 in that transcript (not skipping of 2 exons), and '2-3' denotes skipping of exons 2 and 3.</p>
</div>

<button class="accordion">How is cryptic distance calculated?</button>
<div class="panel">
  <p>We report the maximum number of reads observed in any one sample for each splicing event, across samples from GTEx. We report only for GTEx samples as these are all healthy samples unlike SRA/Intropolis which may have misleading outliers. If the max reads of annotated splicing is low (&#60;250) we recommend not excluding any events missing from 300K-RNA as it could just be due to poor read depth.</p>
</div>

<button class="accordion">How were the default settings chosen? (top 4 events, skipping &#60; 2 exons and cryptics within +/- 600 nt)?</button>
<div class="panel">
  <p>The default settings were chosen as they gave a Sensitivity of 90% and PPV of 31% for predicting exon skipping and cryptic activation events in <a href="https://doi.org/10.1038/s41588-022-01293-8" target="_blank">Dawes et al. 2022</a>.
Users are free to customise settings for their own purposes, however we recommend these thresholds for clinical use.</p>
</div>

<button class="accordion">Can I export the data?</button>
<div class="panel">
  <p>To export the table you've generated, click either copy or download and select file type. If you are planning to open the file in Excel, we recommend the 'Excel' format so that the 'Exons skipped' column doesn't get converted to dates.</p>
  <p>To download the 300K-RNA database see <a href="https://storage.googleapis.com/misspl-db-data/misspl_events_300k_hg38.sql.gz" target="_blank">https://storage.googleapis.com/misspl-db-data/misspl_events_300k_hg38.sql.gz</a>, and to download the 40K-RNA database see <a href="https://storage.googleapis.com/misspl-db-data/misspl_events_40k_hg19.sql.gz" target="_blank">https://storage.googleapis.com/misspl-db-data/misspl_events_40k_hg19.sql.gz</a>.</p>
</div>

<button class="accordion">How do I open IGV links?</button>
<div class="panel">
  <p>In your IGV desktop app, go to Preferences > Advanced &#38; check Enable Port with Port number 60151. You may need to restart IGV after enabling</p>
  <p>IGV must be open when the link is clicked for it to work, and will open so that the donor and acceptor used in the splice junction are at both edges of the screen.</p>
</div>

<button class="accordion">Why does SpliceVault freeze after a few minutes?</button>
<div class="panel">
  <p>We’re hosting SpliceVault on shinyapps.io; which pauses the website if not used for 10 minutes or more (to save computing costs). You can simply click “reload” and continue your analysis.</p>
</div>

<button class="accordion">How can I use SpliceVault through Ensembl VEP?</button>
<div class="panel">
  <p>SpliceVault can now also be used to batch process variants through Ensembl's VEP. Please refer to <a href="https://github.com/kidsneuro-lab/SpliceVault/wiki/Use-of-SpliceVault-through-Ensembl-VEP" target="_blank">Usage of SpliceVault through VEP</a> for more information.</p>
</div>

<button class="accordion">How do I cite SpliceVault?</button>
<div class="panel">
  <p><a href="https://doi.org/10.1038/s41588-022-01293-8" target="_blank">Dawes et al. 2022</a></p>
</div>

<script>
var acc = document.getElementsByClassName("accordion");
var i;

for (i = 0; i < acc.length; i++) {
  acc[i].addEventListener("click", function() {
    this.classList.toggle("active-question");
    var panel = this.nextElementSibling;
    if (panel.style.display === "block") {
      panel.style.display = "none";
    } else {
      panel.style.display = "block";
    }
  });
}
</script>

</body>
</html>
